api/src/main/java/ai/djl/training/listener/EarlyStoppingListener.java

```diff
@@ -0,0 +1,175 @@
+        this.objectiveSuccess = objectiveSuccess;
+        this.minEpochs = minEpochs;
+        this.maxMinutes = maxMinutes;
+        this.earlyStopPctImprovement = earlyStopPctImprovement;
+        this.epochPatience = earlyStopPatience;
+    }
+
+    @Override
+    public void onEpoch(Trainer trainer) {
+        int currentEpoch = trainer.getTrainingResult().getEpoch();
+        // stopping criteria
+        final double loss = getLoss(trainer);
+        if (loss < objectiveSuccess) {
+            throw new EarlyStoppedException(currentEpoch, String.format("validation loss %s < objectiveSuccess %s", loss, objectiveSuccess));
+        }
+        if (currentEpoch >= minEpochs) {
+            double elapsedMinutes = (System.currentTimeMillis() - startTimeMills) / 60_000.0;
+            if (elapsedMinutes >= maxMinutes) {
+                throw new EarlyStoppedException(currentEpoch, String.format("Early stopping training: %.1f minutes elapsed >= %s maxMinutes", elapsedMinutes, maxMinutes));
+            }
+            // consider early stopping?
+            if (Double.isFinite(prevLoss)) {
+                double goalImprovement = prevLoss * (100 - earlyStopPctImprovement) / 100.0;
+                boolean improved = loss <= goalImprovement; // false if any NANs
+                if (improved) {
+                    numberOfEpochsWithoutImprovements = 0;
+                } else {
+                    numberOfEpochsWithoutImprovements++;
+                    if (numberOfEpochsWithoutImprovements >= epochPatience) {
+                        throw new EarlyStoppedException(currentEpoch, String.format("failed to achieve %s%% improvement %s times in a row",
+                                earlyStopPctImprovement, epochPatience));
+                    }
+                }
+            }
+        }
+        if (Double.isFinite(loss)) {
+            prevLoss = loss;
+        }
+    }
+
+    private static double getLoss(Trainer trainer) {
+        Float vLoss = trainer.getTrainingResult().getValidateLoss();
+        if (vLoss != null) {
+            return vLoss;
+        }
+        Float tLoss = trainer.getTrainingResult().getTrainLoss();
+        if (tLoss == null) {
+            return Double.NaN;
+        }
+        return tLoss;
+    }
+
+    @Override
+    public void onTrainingBatch(Trainer trainer, BatchData batchData) {
+        // do nothing
+    }
+
+    @Override
+    public void onValidationBatch(Trainer trainer, BatchData batchData) {
+        // do nothing
+    }
+
+    @Override
+    public void onTrainingBegin(Trainer trainer) {
+        this.startTimeMills = System.currentTimeMillis();
+        this.prevLoss = Double.NaN;
+        this.numberOfEpochsWithoutImprovements = 0;
+    }
+
+    @Override
+    public void onTrainingEnd(Trainer trainer) {
+        // do nothing
+    }
+
+    public EarlyStoppingListener setMinEpochs(int minEpochs) {
```

Comment:
Instead of adding setters on the listener, you may want to make a builder instead. I don't see a reason that users would be modifying this listener while training. Instead, having a builder will make it easier to work with the number of possible arguments where some of them are defaults and others are specified. 

We typically have our builders called by `EarlyStoppingLister.builder().setters...().build()` with arguments as either `setXXX()` or `optXXX()` if required or optional

Reply:
Introduced a build, with all optional parameters, as you can pick whatever you want.